Drop USB interface before reset

dingari opened the pull request:
This is required on (at least) macOS, as the `usb_reset` method will fail while there are interfaces claimed.

This should fix #6 

Relies on https://github.com/dfu-rs/dfu-core/pull/31

cecton commented on flake.nix:
```diff
@@ -0,0 +1,21 @@
+          ];
+        };
+      });
+}
```
Ah good I'm adding those everywhere ^_^

cecton commented on src/lib.rs:
```diff
@@ -199,20 +210,26 @@ impl DfuAsyncIo for DfuNusb {
 
-    async fn usb_reset(&self) -> Result<Self::Reset, Self::Error> {
+    async fn usb_reset(&mut self) -> Result<Self::Reset, Self::Error> {
+        self.interface.take().map(|iface| drop(iface));
```
That's a long way to write `self.interface = None` xD I don't mind though

Is there any reason for doing this that way? It's perfectly okay if it's just a personal preference! I'm just curious.

dingari replied:
Haha wow. That's a brainfart xD

I'll change it

cecton commented:
@sjoerdsimons is this PR okay for you?

Notable changes:
- added nix files (useful for some, doesn't hurt)
- adapted the code to drop the interface after the `usb_reset()`

sjoerdsimons commented:
I'd rather not include the nix files; I don't use nix, so that'll just go unmaintained.. I don't think crates should have distro specific configuration in them as that's a very slippery slope.

As it stands this commit still makes dfu-nusb use a forked dfu-core, so that should be dropped  before merging. Also should do a semver bump as the api is changed.

The code i'd need to further review (as always end of the year is hectic, so i'm slow).. But from a quick look the change to the `usb_reset` signature seems awkward, simply taking self would end up a lot more elegant (and a much smaller delta).

sjoerdsimons commented:
@cecton  btw just quickly looking at dfu-core ; looks like 0.9.2 is a semver break (changing the type of arguments); That should have been 0.10 instead

dingari commented:
Totally understand wrt to the nix devshell. I always meant to remove them before PR-ing. I'll fixup the commit and re-push.

> As it stands this commit still makes dfu-nusb use a forked dfu-core, so that should be dropped before merging. 

Oh, yeah! Totally. I'll wait for a new `dfu-core` semver bump and fix that. 

> Also should do a semver bump as the api is changed.

I figured a maintainer would do that before issuing a new release?

dingari commented:
> But from a quick look the change to the usb_reset signature seems awkward, simply taking self would end up a lot more elegant (and a much smaller delta).

Sure, I'll be happy to make that change to `dfu-core` (consuming self), and updating this PR to reflect that.

dingari commented:
@sjoerdsimons making `usb_reset` consume `self` just shifts the partial-move into the `DfuASync` / `DfuSync` impls. As `download` calls `usb_reset`.  

```
error[E0507]: cannot move out of `self.io` which is behind a mutable reference
   --> /Users/genki/development/dfu-core/src/asynchronous.rs:239:21
    |
239 |                     self.io.usb_reset().await?;
    |                     ^^^^^^^ ----------- `self.io` moved due to this method call
    |                     |
    |                     move occurs because `self.io` has type `IO`, which does not implement the `Copy` trait
    |
note: `DfuAsyncIo::usb_reset` takes ownership of the receiver `self`, which moves `self.io`
   --> /Users/genki/development/dfu-core/src/asynchronous.rs:40:18
    |
40  |     fn usb_reset(self) -> impl Future<Output = Result<Self::Reset, Self::Error>> + Send;
    |                  ^^^^
help: if `IO` implemented `Clone`, you could clone the value
   --> /Users/genki/development/dfu-core/src/asynchronous.rs:157:6
    |
157 | impl<IO, E> DfuASync<IO, E>
    |      ^^ consider constraining this type parameter with `Clone`
...
239 |                     self.io.usb_reset().await?;
    |                     ------- you could clone this value

error[E0507]: cannot move out of `self.io` which is behind a mutable reference
   --> /Users/genki/development/dfu-core/src/sync.rs:173:21
    |
173 |                     self.io.usb_reset()?;
    |                     ^^^^^^^ ----------- `self.io` moved due to this method call
    |                     |
    |                     move occurs because `self.io` has type `IO`, which does not implement the `Copy` trait
    |
note: `DfuIo::usb_reset` takes ownership of the receiver `self`, which moves `self.io`
   --> /Users/genki/development/dfu-core/src/lib.rs:107:18
    |
107 |     fn usb_reset(self) -> Result<Self::Reset, Self::Error>;
    |                  ^^^^
help: if `IO` implemented `Clone`, you could clone the value
   --> /Users/genki/development/dfu-core/src/sync.rs:95:6
    |
95  | impl<IO, E> DfuSync<IO, E>
    |      ^^ consider constraining this type parameter with `Clone`
...
173 |                     self.io.usb_reset()?;
    |                     ------- you could clone this value

For more information about this error, try `rustc --explain E0507`.
error: could not compile `dfu-core` (lib) due to 2 previous errors
```

So, honestly, I think the way this PR does it is just OK. I could move the interface-unwrapping in a fn to avoid code duplication.

dingari commented on Cargo.toml:
```diff
@@ -28,3 +28,6 @@ futures = { version = "0.3.31", features = ["io-compat"] }
 indicatif = { version = "0.17.8", features = [ "tokio" ] }
+
+[patch.crates-io]
+dfu-core = { git = "https://github.com/genkiinstruments/dfu-core", rev = "67dcd21613778a461063ab014c8544411c586ada" }
```
Remove this patch once a compatible `dfu-core` released.

cecton commented:
Oops you're right

It's yanked and I released 0.10.

cecton commented:
> @sjoerdsimons making `usb_reset` consume `self` just shifts the partial-move into the `DfuASync` / `DfuSync` impls. As `download` calls `usb_reset`.

Ah yes that was the issue. Using a state machine has its limitations. The IO object must stay alive.

cecton commented:
Seems all good to me. Only waiting on @sjoerdsimons feedback :innocent:

cecton commented:
Hi @sjoerdsimons, I wanted to check in on this PR -- I know everyone's busy and things slip through. Here's where things stand since your last review:

- Nix files have been removed
- The dependency on a forked `dfu-core` has been dropped; this now tracks the published `0.10`
- `usb_reset` takes `&mut self` as discussed

The only remaining item is a semver bump to `0.2.0` (breaking API change on `usb_reset`), which I'll handle before merging.

If I don't hear any concerns in the next two weeks, I'll go ahead and merge. This is a real bug on macOS and it'd be great to get it shipped.

sjoerdsimons commented:
fwiw Looking at this ; sorry for taking so long, start of this year has been a bit mad

sjoerdsimons committed:
chore: update dfu-core dependency

Move to dfu-core #11; Primary change is that usb_reset now takes
ownership of the i/o trait to more complete de-initializing before
resetting the device. This is in particular needed on Mac OS where the
device can only be reset if no interfaces are claimed.

Also change the usb_reset implementation to actually do the above.

Fixes: #6
Co-developed-by: dingari <daniel@genkiinstruments.com>
Co-developed-by: Cecile Tonglet <cecile.tonglet@cecton.com>

sjoerdsimons commented:
Having some fun with this; Put together https://github.com/dfu-rs/dfu-core/pull/32  -- The ownership issue really isn't an issue for the state machine (that doesn't care), it's about the download helpers on the Dfu helpers.

Doing this makes the equivalent of this PR quite simple: https://github.com/dfu-rs/dfu-nusb/compare/main...usb-reset-owned#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759 -- which is nice.

If does cause a bit more churn for the users; But it does make for better semantics imho as the I/O object will be a dud after download has done a usb reset anyway, so any usage at that point will just cause errors.

@dingari again sorry for taking so long; The one thing i couldn't test here is with the change above it would fix it for Mac OS; Mostly as i didn't verify yet if decomposing the self object in reset drops the nusb interface in time.

dingari commented:
@sjoerdsimons no worries. I can see if the decomposition does drop the nusb interface in time on macOS tomorrow.

As discussed in the wayback, I do agree that consuming `self` is semantically more correct.

sjoerdsimons commented:
Fwiw just destructuring isn't enough see: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=d9e0a45a085db5d6ab937d72578dd7f5 

So we'll have to be more explicit about it; Lets see what @cecton things about the dfu-core PR and then we can go from there :)

dingari commented:
Yeah, fwiw an explicit call to `drop(self.interface)`  before `self.device.reset()`is very clear and intentional. And less prone to regresssion in the future.

cecton commented:
I understand now and I agree consuming self would be much better.

I reviewed the dfu-core PR and added some additional changes which I believe will be more useful for embedded developers or people building tools for very specific devices. @sjoerdsimons can you review?

I applied the change for `drop(self.interface)`.

@dingari could you reset your branch to `usb-reset-owned`? I would like to keep you as contributor and author if that's okay for you.

dingari commented:
@cecton Sure! I've done that just now :)

cecton commented on Cargo.toml:
```diff
@@ -15,7 +15,7 @@ tokio = ["dep:tokio"]
 
 [dependencies]
-dfu-core = { version = "0.8.0", features = ["async"] }
+dfu-core = { path = "../dfu-core", version = "0.11.0-rc0", features = ["async"] }
```
Not sure what happened here but I think it was a mistake on my side :grin: 

Version of dfu-core is now 0.11.0

dingari replied:
Done :)

cecton reviewed (approved):
Ok dfu-core 0.11.0 is released. Can you update the dependency?

cecton commented:
I think I got a bit ahead of myself. I'll let @sjoerdsimons handle the merge and the release. Let me know if you want me to update/improve things (README, badges, changelog, msrv, ...) right before the release (on main branch).

dingari commented:
Yeah, seems like maybe the runner toolchain needs to be updated?

sjoerdsimons commented:
Heh yes this got a bit of a mess; i force pushed a squashed commit to the usb-reset-owned as it's hard to seperate at this point, keeping @dingari and @cecton as co-developed-by  -- It seems like this PR doens't allow pushes from maintainers otherwise i had it pushed it here for clarity.

@cecton i ddi change the comment in the usb_reset functions a bit; I'm guessing those were AI written as they were overly verbose and slightly besides the point/misleading :)

Will look at the CI failure next ; @dingari just to confirm, with these changes things work fine on mac os for you now right ? (i can't test that here)

sjoerdsimons committed:
chore: ci: bump ci toolchain

rustc 1.78 was released over 2 years ago. Time to bump to a newer
toolchain as CI started to fail due to dependencies wanting rust
editition 2024.

No changes in MSRV done

dingari commented:
@sjoerdsimons Yes works a treat on macOS on `usb-reset-owned` 🎉

sjoerdsimons commented:
@dingari Cool I'll merge it tonight. If you're ok, if you could reset your branch/this MR on that branch i can merge it as this MR so we get a reference to the history (as said i can't push to it for some reason).. Otherwise i'll just merge through a seperate MR

dingari commented:
@sjoerdsimons Done

I don't know why but I don't have the "maintainers can edit..." check in the Github UI - maybe it's because the fork is under my organization?

sjoerdsimons commented:
> @sjoerdsimons Done
> 
> I don't know why but I don't have the "maintainers can edit..." check in the Github UI - maybe it's because the fork is under my organization?

fwiw it should show up in your PR request; https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork -- and i can also a ticky box for it on the right-hand bar on an open PR (below participants).  the docs don't mention anything about organisation vs. personal repos, but show knows.

In any case thanks for doing the reset and the patience!

cecton commented:
> maybe it's because the fork is under my organization?

Yes that's the reason. I had the issue too
